### PR TITLE
Revise validation errors

### PR DESCRIPTION
**Pull request by ak88**
## Changes

This will set error messages primarily for engine RPC endpoints, which is used in some hive test to test for correctness.

Format for error messages is two segments divided with ':'
- first segment is machine parseable and can be used for verification
- second part is a human readable error description
Example: 
`"InvalidAncestor: No valid ancestors could be found."`
`"InvalidBaseFeePerGas: Does not match calculated."`

This is meant to serve as a temporary solution until there is a common standard for Json RPC error codes.

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [ ] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Consensus/Processing/BlockProcessor.cs [hunk 1]
```diff
@@ -213,11 +213,11 @@ private void RestoreBranch(Hash256 branchingPointStateRoot)
...
+        if (!options.ContainsFlag(ProcessingOptions.NoValidation) && !_blockValidator.ValidateProcessedBlock(block, receipts, suggestedBlock, out string? error))
         {
-            if (_logger.IsWarn) _logger.Warn($"Processed block is not valid {suggestedBlock.ToString(Block.Format.FullHashAndNumber)}");
-            if (_logger.IsWarn) _logger.Warn($"Suggested block TD: {suggestedBlock.TotalDifficulty}, Suggested block IsPostMerge {suggestedBlock.IsPostMerge}, Block TD: {block.TotalDifficulty}, Block IsPostMerge {block.IsPostMerge}");
-            throw new InvalidBlockException(suggestedBlock, "invalid hash after block processing");
+            if (_logger.IsError) _logger.Error($"Processed block is not valid {suggestedBlock.ToString(Block.Format.FullHashAndNumber)}");
+            if (_logger.IsError) _logger.Error($"Suggested block TD: {suggestedBlock.TotalDifficulty}, Suggested block IsPostMerge {suggestedBlock.IsPostMerge}, Block TD: {block.TotalDifficulty}, Block IsPostMerge {block.IsPostMerge}");
+            throw new InvalidBlockException(suggestedBlock, error);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/Processing/BlockProcessor.cs:220]: Are we sure we want to do Error log here? Sometimes blocks are correctly identified as bad blocks.

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.Consensus/Processing/BlockProcessor.cs (hunk 1 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Consensus/Processing/BlockProcessor.cs:220]: Also sometimes bad blocks are proposed, isn't a terrible thing

File: src/Nethermind/Nethermind.Consensus/Validators/TxValidator.cs [hunk 2]
```diff
@@ -106,54 +149,119 @@ private bool ValidateSignature(Transaction tx, IReleaseSpec spec)
...
+                error = TxErrorMessages.BlobTxMissingBlobVersionedHashes;
+                return false;
             }
 
-            if (transaction.To is null ||
-                transaction.MaxFeePerBlobGas is null ||
-                transaction.BlobVersionedHashes is null ||
-                BlobGasCalculator.CalculateBlobGas(transaction.BlobVersionedHashes!.Length) > Eip4844Constants.MaxBlobGasPerTransaction ||
-                transaction.BlobVersionedHashes!.Length < Eip4844Constants.MinBlobsPerTransaction)
+            var totalDataGas = BlobGasCalculator.CalculateBlobGas(transaction.BlobVersionedHashes!.Length);
```
Comment (flcl42) [inline src/Nethermind/Nethermind.Consensus/Validators/TxValidator.cs:197]: var

File: src/Nethermind/Nethermind.Consensus/Validators/TxValidator.cs [hunk 3]
```diff
@@ -106,54 +149,119 @@ private bool ValidateSignature(Transaction tx, IReleaseSpec spec)
...
+                error = TxErrorMessages.BlobTxGasLimitExceeded;
+                return false;
+            }
+            if (transaction.BlobVersionedHashes!.Length < Eip4844Constants.MinBlobsPerTransaction)
+            {
+                error = TxErrorMessages.BlobTxMissingBlobs;
                 return false;
             }
 
             int blobCount = transaction.BlobVersionedHashes.Length;
```
Comment (flcl42) [inline src/Nethermind/Nethermind.Consensus/Validators/TxValidator.cs:209]: May be even useful above on L197, L203

Comment (ak88) [reply src/Nethermind/Nethermind.Consensus/Validators/TxValidator.cs:209]: I am not sure what you mean here?

File: src/Nethermind/Nethermind.Consensus/Validators/TxValidator.cs (hunk 2 above)
Comment (ak88) [reply src/Nethermind/Nethermind.Consensus/Validators/TxValidator.cs]: fixed

File: src/Nethermind/Nethermind.Consensus/Processing/BlockProcessor.cs (hunk 1 above)
Comment (ak88) [reply src/Nethermind/Nethermind.Consensus/Processing/BlockProcessor.cs]: Changed to Warning and included the error in the log statement

File: src/Nethermind/Nethermind.Blockchain.Test/Validators/BlockValidatorTests.cs [hunk 4]
```diff
@@ -88,5 +94,120 @@ public void ValidateBodyAgainstHeader_BlockHasInvalidWithdrawalsRoot_ReturnsFals
...
+            Assert.That(error, Does.StartWith("InvalidStateRoot"));
+        }
+
+        [Test]
+        public void ValidateSuggestedBlock_UncleHashIsWrong_ErrorIsSet()
+        {
+            TxValidator txValidator = new(TestBlockchainIds.ChainId);
+            ISpecProvider specProvider = Substitute.For<ISpecProvider>();
+            BlockValidator sut = new(txValidator, Always.Valid, Always.Valid, specProvider, LimboLogs.Instance);
+            Block suggestedBlock = Build.A.Block.WithHeader(Build.A.BlockHeader.WithUnclesHash(Keccak.Zero).TestObject).TestObject;
```
[the remark this answers is not in the document]
Comment (ak88) [reply src/Nethermind/Nethermind.Blockchain.Test/Validators/BlockValidatorTests.cs]: I converted to `TestCaseSource`, but i ran across an [annoying problem here](https://github.com/NethermindEth/nethermind/blob/86e5326f132f9361154ea12e1191dd698f2e964c/src/Nethermind/Nethermind.Core.Test/Builders/TransactionBuilder.cs#L195)

Kzg is not initialized when `TestCaseSource` is run, which is difficult to catch. Maybe we should just throw iif not initialized?

File: src/Nethermind/Nethermind.Evm.Test/TransactionSubstateTests.cs [hunk 5]
```diff
@@ -136,7 +136,7 @@ private static IEnumerable<(byte[], string)> PanicFunctionTestCases()
                 "Reverted unknown panic code (0xff)");
 
             // Invalid case
-            yield return (new byte[] { 0xf0, 0x28, 0x8c, 0x28 }, "Reverted 0xf0288c28");
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Evm.Test/TransactionSubstateTests.cs:139]: Why changing this? Needs to test invalid Utf8 and make sure it gets output as hex

File: src/Nethermind/Nethermind.Consensus/Messages/TxErrorMessages.cs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    public const string TxMissingTo =
+        "TxMissingTo: Must be set.";
+
+    public const string BlobTxMissingMaxFeePerBlobGas =
+        "BlobTxMissingMaxFeePerBlobGas: Must be set.";
+
+    public const string BlobTxMissingBlobVersionedHashes =
+        "BlobTxMissingBlobVersionedHashes: Must be set.";
+
+    public static string BlobTxGasLimitExceeded =>
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Consensus/Messages/TxErrorMessages.cs:49]:
```suggestion
    public static readonly string BlobTxGasLimitExceeded =
```

File: src/Nethermind/Nethermind.Consensus/Messages/TxErrorMessages.cs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        $"BlobTxGasLimitExceeded: Transaction exceeded {Eip4844Constants.MaxBlobGasPerTransaction}.";
+
+    public const string BlobTxMissingBlobs =
+        "BlobTxMissingBlobs: Blob transaction must have blobs.";
+
+    public const string MissingBlobVersionedHash =
+        "MissingBlobVersionedHash: Must be set.";
+
+    public static string InvalidBlobVersionedHashSize =>
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Consensus/Messages/TxErrorMessages.cs:58]:
```suggestion
    public static readonly string InvalidBlobVersionedHashSize =
```

File: src/Nethermind/Nethermind.Consensus/Messages/TxErrorMessages.cs [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        $"InvalidBlobVersionedHashSize: Cannot exceed {KzgPolynomialCommitments.BytesPerBlobVersionedHash}.";
+
+    public const string InvalidBlobVersionedHashVersion =
+        "InvalidBlobVersionedHashVersion: Blob version not supported.";
+
+    public static string ExceededBlobSize =>
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Consensus/Messages/TxErrorMessages.cs:64]:
```suggestion
    public static readonly string ExceededBlobSize =
```

File: src/Nethermind/Nethermind.Consensus/Messages/TxErrorMessages.cs [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        $"ExceededBlobSize: Cannot be more than {Ckzg.Ckzg.BytesPerBlob}.";
+
+    public static string ExceededBlobCommitmentSize =>
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Consensus/Messages/TxErrorMessages.cs:67]:
```suggestion
    public static readonly string ExceededBlobCommitmentSize =
```

File: src/Nethermind/Nethermind.Consensus/Messages/BlockErrorMessages.cs [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+        "MissingExcessBlobGas: Must be set in header.";
+
+    public static string InvalidTxInBlock(int i) =>
+        $"InvalidTxInBlock: Tx at index {i} in body.";
+
+    public const string HeaderGasUsedMismatch =
+        "HeaderGasUsedMismatch: Gas used in header does not match calculated.";
+
+    //Block's blob gas used in header is above the limit.
+    public static string BlobGasUsedAboveBlockLimit() =>
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Consensus/Messages/BlockErrorMessages.cs:98]:
```suggestion
    public static readonly string BlobGasUsedAboveBlockLimit =
```

**Review by benaadams — APPROVED** [review]
Are some error messages that don't need to be allocated each time they are used and can be made `readonly static`s and assigned `=` rather than be a function call `=>` that allocates each time

File: src/Nethermind/Nethermind.Consensus/Validators/TxValidator.cs (hunk 3 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Consensus/Validators/TxValidator.cs:209]: this property is accessed earlier, lets put it in variable earlier

File: src/Nethermind/Nethermind.Evm.Test/TransactionSubstateTests.cs (hunk 5 above)
Comment (ak88) [reply src/Nethermind/Nethermind.Evm.Test/TransactionSubstateTests.cs:139]: > Why changing this? Needs to test invalid Utf8 and make sure it gets output as hex

Comes from https://github.com/NethermindEth/nethermind-security/commit/46de1e5cbc872af0a2b85781e506d16963ce3dc3 that was merged into this branch

Comment (benaadams) [reply src/Nethermind/Nethermind.Evm.Test/TransactionSubstateTests.cs:139]: Is already in master, then this test was fixed, so please revert this line. Already test valid cases

**Review by benaadams — APPROVED** [review]

**Pull request merged**